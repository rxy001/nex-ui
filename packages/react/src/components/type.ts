export type ClassNames = string | string[]

export type ComponentClasses<OwnerState, T extends string> = Partial<
  Record<T, ClassNames | ((ownerState: OwnerState) => ClassNames)>
>
