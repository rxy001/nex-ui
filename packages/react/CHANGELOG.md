# @nex-ui/react

## 0.0.15

### Patch Changes

- fdd4483: Release
- Updated dependencies [fdd4483]
  - @nex-ui/styled@0.0.15
  - @nex-ui/system@0.0.15
  - @nex-ui/hooks@0.0.15
  - @nex-ui/icons@0.0.15
  - @nex-ui/utils@0.0.15

## 0.0.14

### Patch Changes

- ca25983: Release
- Updated dependencies [ca25983]
  - @nex-ui/styled@0.0.14
  - @nex-ui/system@0.0.14
  - @nex-ui/hooks@0.0.14
  - @nex-ui/icons@0.0.14
  - @nex-ui/utils@0.0.14

## 0.0.13

### Patch Changes

- 1b97788: Release
- Updated dependencies [1b97788]
  - @nex-ui/styled@0.0.13
  - @nex-ui/system@0.0.13
  - @nex-ui/hooks@0.0.13
  - @nex-ui/icons@0.0.13
  - @nex-ui/utils@0.0.13

## 0.0.12

### Patch Changes

- 664eb21: Test

## 0.0.11

### Patch Changes

- 259356e: Test
- Updated dependencies [259356e]
  - @nex-ui/styled@0.0.11
  - @nex-ui/system@0.0.11
  - @nex-ui/hooks@0.0.11
  - @nex-ui/icons@0.0.11
  - @nex-ui/utils@0.0.10

## 0.0.10

### Patch Changes

- 5a229bd: release
- Updated dependencies [5a229bd]
  - @nex-ui/styled@0.0.10
  - @nex-ui/system@0.0.10
  - @nex-ui/icons@0.0.10

## 0.0.9

### Patch Changes

- 4a5c359: release
- Updated dependencies [4a5c359]
  - @nex-ui/styled@0.0.9
  - @nex-ui/system@0.0.9
  - @nex-ui/hooks@0.0.9
  - @nex-ui/icons@0.0.9
  - @nex-ui/utils@0.0.9

## 0.0.8

### Patch Changes

- c8c1956: release
- Updated dependencies [c8c1956]
  - @nex-ui/styled@0.0.8
  - @nex-ui/system@0.0.8
  - @nex-ui/hooks@0.0.8
  - @nex-ui/icons@0.0.8
  - @nex-ui/utils@0.0.8

## 0.0.7

### Patch Changes

- 2618e1a: release
- Updated dependencies [2618e1a]
  - @nex-ui/styled@0.0.7
  - @nex-ui/system@0.0.7
  - @nex-ui/hooks@0.0.7
  - @nex-ui/icons@0.0.7
  - @nex-ui/utils@0.0.7

## 0.0.6

### Patch Changes

- 87191c1: release
- Updated dependencies [87191c1]
  - @nex-ui/styled@0.0.6
  - @nex-ui/system@0.0.6
  - @nex-ui/hooks@0.0.6
  - @nex-ui/icons@0.0.6
  - @nex-ui/utils@0.0.6

## 0.0.5

### Patch Changes

- 7631577: release
- Updated dependencies [7631577]
  - @nex-ui/styled@0.0.5
  - @nex-ui/system@0.0.5
  - @nex-ui/hooks@0.0.5
  - @nex-ui/icons@0.0.5
  - @nex-ui/utils@0.0.5

## 0.0.4

### Patch Changes

- 70cb478: v0.0.4
- Updated dependencies [70cb478]
  - @nex-ui/styled@0.0.4
  - @nex-ui/system@0.0.4
  - @nex-ui/hooks@0.0.4
  - @nex-ui/icons@0.0.4
  - @nex-ui/utils@0.0.4
