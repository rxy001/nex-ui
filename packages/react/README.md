<div style="background:white;">
  <p align="center">
    <img src='../../assets/logo.png' style="width: 150px; height: 150px;" />
  </p>
  <h1 align="center">Nex UI</h1>
  <p align="center">ðŸŽ‰ A beautiful, modern, and reliable React component library.</p>
  <p align="center"><b>Theme Customizable, Light & Dark UI, Unique DX, Written In TypeScript</b></p>
</div>

## Documentation

Visit https://nex-ui-docs.vercel.app to view the full documentation.

## Getting Started

Visit https://nex-ui-docs.vercel.app/docs/getting-started to get started with Nex UI.

## Storybook

Visit https://nex-ui-storybook.vercel.app to view the storybook for all components.

## Contributing

Contributions are always welcome!

See [CONTRIBUTING.md](/CONTRIBUTING.md) for ways to get started.

Please adhere to this project's [CODE_OF_CONDUCT](/CODE_OF_CONDUCT.md).

## License

[MIT](https://choosealicense.com/licenses/mit/)
