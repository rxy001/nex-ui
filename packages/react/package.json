{
  "name": "@nex-ui/react",
  "version": "0.0.12",
  "author": "X1ng Yu",
  "description": "ðŸŽ‰ A beautiful, modern, and reliable React component library.",
  "homepage": "https://nex-ui-docs.vercel.app/",
  "private": false,
  "keywords": [
    "react",
    "react-component",
    "nui",
    "nex-ui"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/rxy001/nex-ui.git",
    "directory": "packages/react"
  },
  "license": "MIT",
  "scripts": {
    "build": "tsx ../../scripts/build.ts --dts",
    "build:fast": "tsx ../../scripts/build.ts",
    "typecheck": "tsc --noEmit",
    "prepack": "clean-package",
    "postpack": "clean-package restore"
  },
  "type": "module",
  "files": [
    "dist"
  ],
  "main": "src/index.ts",
  "devDependencies": {
    "@storybook/react": "8.6.12"
  },
  "dependencies": {
    "@nex-ui/hooks": "workspace:*",
    "@nex-ui/icons": "workspace:*",
    "@nex-ui/styled": "workspace:*",
    "@nex-ui/system": "workspace:*",
    "@nex-ui/utils": "workspace:*",
    "clsx": "2.1.1"
  },
  "peerDependencies": {
    "@emotion/react": ">=11",
    "motion": "^12.7.2",
    "react": ">=19",
    "react-dom": ">=19"
  },
  "sideEffects": false,
  "clean-package": "../../clean-package.config.json"
}
