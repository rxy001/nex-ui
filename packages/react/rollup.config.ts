import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { defineConfig } from 'rollup'
import typescript from '@rollup/plugin-typescript'
import del from 'rollup-plugin-delete'
import dts from 'rollup-plugin-dts'
import { nexUIRollupPlugin } from '@nex-ui/plugins'

const dirname = fileURLToPath(new URL('./', import.meta.url))

export default defineConfig([
  {
    plugins: [
      del({
        targets: './dist/*',
      }),
      typescript({
        tsconfig: path.resolve(dirname, 'tsconfig.json'),
      }),
      nexUIRollupPlugin(),
    ],
    input: './src/index.ts',
    external: [
      'react',
      'react-dom',
      'react/jsx-runtime',
      'react-dom/client',
      '@nex-ui/utils',
      '@nex-ui/css-system',
      'classnames',
      'framer-motion',
      '@iconify/react',
      '@vanilla-extract/dynamic',
      '@vanilla-extract/css',
      '@vanilla-extract/css/functionSerializer',
    ],
    output: [
      {
        preserveModules: true,
        format: 'es',
        dir: './dist/es',
        assetFileNames({ name }) {
          return name?.replace(/^src\//, '') ?? ''
        },
      },
      {
        preserveModules: true,
        format: 'cjs',
        dir: './dist/lib',
        assetFileNames({ name }) {
          return name?.replace(/^src\//, '') ?? ''
        },
      },
    ],
  },
  {
    plugins: [dts()],
    input: './src/index.ts',
    output: {
      dir: './dist/types',
      format: 'es',
      preserveModules: true,
    },
  },
])
