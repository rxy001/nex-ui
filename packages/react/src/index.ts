import './types/declare'

export * from './components'
export * from './theme'

export type { Theme } from './types/theme'
export type { ComponentsTheme } from './types/componentsTheme'

export { useColorScheme } from '@nex-ui/system'
export { nex, styled } from '@nex-ui/styled'
