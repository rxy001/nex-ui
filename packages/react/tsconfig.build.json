{
  "extends": "../../tsconfig.json",
  "include": ["./src"],
  "exclude": ["./src/**/__tests__/**/*", "./src/**/__stories__/**/*"]
}
