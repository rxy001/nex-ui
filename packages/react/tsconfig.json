{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "~/*": ["../../*"],
      "@theme": ["./src/theme"]
    }
  },
  "include": ["./src", "./rollup.config.ts"]
}
