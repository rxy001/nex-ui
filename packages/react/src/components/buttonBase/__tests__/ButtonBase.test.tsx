import {
  testComponentStability,
  renderWithNexUIProvider,
  testRefForwarding,
  testVariantDataAttrs,
} from '~/tests/shared'
import { fireEvent } from '@testing-library/react'
import { ButtonBase } from '../index'
import type { ButtonHTMLAttributes } from 'react'

describe('ButtonBase', () => {
  testComponentStability(<ButtonBase />)

  testRefForwarding(<ButtonBase />)

  testVariantDataAttrs(<ButtonBase>Button</ButtonBase>, [
    'disabled',
    [true, false],
  ])

  it('should render with default props', () => {
    const { container } = renderWithNexUIProvider(
      <ButtonBase>Button</ButtonBase>,
    )

    const root = container.firstElementChild

    expect(root).toMatchSnapshot()
  })

  it('should have type="button" by default', () => {
    const { getByRole } = renderWithNexUIProvider(
      <ButtonBase>Button</ButtonBase>,
    )
    expect(getByRole('button')).toHaveAttribute('type', 'button')
  })

  it('should change type when type prop is provided', () => {
    const { getByRole } = renderWithNexUIProvider(
      <ButtonBase type='submit'>Submit</ButtonBase>,
    )
    expect(getByRole('button')).toHaveAttribute('type', 'submit')
  })

  it('should change component type and add accessibility attributes when as prop is provided', () => {
    const { getByRole } = renderWithNexUIProvider(
      <ButtonBase as='span'>Link</ButtonBase>,
    )
    const button = getByRole('button')
    expect(button.tagName).toBe('SPAN')
    expect(button).toHaveAttribute('tabIndex', '0')
    expect(button).not.toHaveAttribute('type')
  })

  it('should not apply role="button" when rendered as a button', () => {
    const { getByRole } = renderWithNexUIProvider(
      <ButtonBase as='button'>Button</ButtonBase>,
    )
    const button = getByRole('button')
    expect(button).not.toHaveAttribute('role', 'button')
  })

  it('should automatically change the button to an anchor element when href is provided', () => {
    const { getByRole } = renderWithNexUIProvider(
      <ButtonBase href='https://example.com'>Link</ButtonBase>,
    )
    const link = getByRole('link')
    expect(link.tagName).toBe('A')
    expect(link).toHaveAttribute('href', 'https://example.com')
    expect(link).not.toHaveAttribute('type')
    expect(link).not.toHaveAttribute('role')
  })

  it('should apply role="button" when anchor is used without href', () => {
    const { getByRole } = renderWithNexUIProvider(
      <ButtonBase as='a'>Anchor</ButtonBase>,
    )
    const anchor = getByRole('button')
    expect(anchor.tagName).toBe('A')
    expect(anchor).not.toHaveAttribute('type')
  })

  it('should not apply role="button" when as prop is custom component', () => {
    const Component = (props: ButtonHTMLAttributes<HTMLButtonElement>) => (
      <button {...props}>Custom</button>
    )

    const { getByRole } = renderWithNexUIProvider(
      <ButtonBase as={Component}>Button</ButtonBase>,
    )

    const button = getByRole('button')
    expect(button).not.toHaveAttribute('role', 'button')
  })

  it('should fire event callbacks', async () => {
    const onClick = jest.fn()
    const onBlur = jest.fn()
    const onFocus = jest.fn()
    const onKeyUp = jest.fn()
    const onKeyDown = jest.fn()
    const onMouseDown = jest.fn()
    const onMouseLeave = jest.fn()
    const onMouseUp = jest.fn()
    const onContextMenu = jest.fn()
    const onTouchStart = jest.fn()
    const onTouchEnd = jest.fn()

    const { getByText } = renderWithNexUIProvider(
      <ButtonBase
        onClick={onClick}
        onBlur={onBlur}
        onFocus={onFocus}
        onKeyUp={onKeyUp}
        onKeyDown={onKeyDown}
        onMouseDown={onMouseDown}
        onMouseLeave={onMouseLeave}
        onMouseUp={onMouseUp}
        onContextMenu={onContextMenu}
        onTouchEnd={onTouchEnd}
        onTouchStart={onTouchStart}
      >
        Hello
      </ButtonBase>,
    )

    const button = getByText('Hello')

    // only run in supported browsers
    if (typeof Touch !== 'undefined') {
      fireEvent.touchStart(button)
      expect(onTouchStart).toHaveBeenCalled()

      fireEvent.touchEnd(button)
      expect(onTouchEnd).toHaveBeenCalled()
    }

    fireEvent.mouseDown(button)
    expect(onMouseDown).toHaveBeenCalled()

    fireEvent.mouseUp(button)
    expect(onMouseUp).toHaveBeenCalled()

    fireEvent.contextMenu(button)
    expect(onContextMenu).toHaveBeenCalled()

    fireEvent.click(button)
    expect(onClick).toHaveBeenCalled()

    fireEvent.focus(button)
    expect(onFocus).toHaveBeenCalled()

    fireEvent.keyDown(button)
    expect(onKeyDown).toHaveBeenCalled()

    fireEvent.keyUp(button)
    expect(onKeyUp).toHaveBeenCalled()

    fireEvent.blur(button)
    expect(onBlur).toHaveBeenCalled()

    fireEvent.mouseLeave(button)
    expect(onMouseLeave).toHaveBeenCalled()
  })

  it('should apply data-focus-visible="true" when keyboard focused', async () => {
    const { getByRole, user } = renderWithNexUIProvider(
      <ButtonBase>Focusable Button</ButtonBase>,
    )

    const button = getByRole('button')

    expect(button).not.toHaveAttribute('data-focus-visible')

    await user.tab()
    expect(document.activeElement).toBe(button)
    expect(button).toHaveAttribute('data-focus-visible', 'true')
  })

  it('should activate button element when Space or Enter is pressed', async () => {
    const onClick = jest.fn()

    const { user, getByRole } = renderWithNexUIProvider(
      <ButtonBase onClick={onClick}>Focusable Button</ButtonBase>,
    )

    const button = getByRole('button')
    await user.tab()
    expect(document.activeElement).toBe(button)

    await user.keyboard(' ')
    expect(onClick).toHaveBeenCalledTimes(1)

    await user.keyboard('{Enter}')
    expect(onClick).toHaveBeenCalledTimes(2)
  })

  it('should activate non-interactive elements when Space or Enter is pressed', async () => {
    const onClick = jest.fn()

    const { user, getByRole } = renderWithNexUIProvider(
      <ButtonBase as='span' onClick={onClick}>
        Focusable Span
      </ButtonBase>,
    )

    const span = getByRole('button')
    await user.tab()
    expect(document.activeElement).toBe(span)

    await user.keyboard(' ')
    expect(onClick).toHaveBeenCalledTimes(1)

    await user.keyboard('{Enter}')
    expect(onClick).toHaveBeenCalledTimes(2)
  })

  it('should prevent repeated click event triggering when the Enter key is pressed', async () => {
    const onClick = jest.fn()

    const { getByRole, user } = renderWithNexUIProvider(
      <ButtonBase onClick={onClick}>Button</ButtonBase>,
    )

    const button = getByRole('button')

    await user.tab()
    expect(document.activeElement).toBe(button)

    await user.keyboard('{Enter>5/}')
    expect(onClick).toHaveBeenCalledTimes(1)
  })

  describe('disabled', () => {
    it('should apply disabled attribute with native button', () => {
      const { getByRole } = renderWithNexUIProvider(
        <ButtonBase disabled>Disabled Button</ButtonBase>,
      )

      const button = getByRole('button')
      expect(button).toBeDisabled()
    })

    it('should apply tabIndex=-1', () => {
      const { getByRole } = renderWithNexUIProvider(
        <ButtonBase disabled>Disabled Button</ButtonBase>,
      )

      const button = getByRole('button')
      expect(button).toHaveAttribute('tabIndex', '-1')
    })

    it('should not use aria-disabled with native button', () => {
      const { getByRole } = renderWithNexUIProvider(
        <ButtonBase disabled>Disabled Button</ButtonBase>,
      )

      const button = getByRole('button')
      expect(button).not.toHaveAttribute('aria-disabled')
    })

    it('should use aria-disabled with non-button', () => {
      const { getByRole } = renderWithNexUIProvider(
        <ButtonBase as='span' disabled>
          Disabled Span
        </ButtonBase>,
      )

      const span = getByRole('button')
      expect(span).toHaveAttribute('aria-disabled', 'true')
    })

    it('should not fire event callbacks', () => {
      const onClick = jest.fn()
      const onKeyUp = jest.fn()
      const onKeyDown = jest.fn()

      const { getByRole } = renderWithNexUIProvider(
        <ButtonBase
          as='span'
          disabled
          onClick={onClick}
          onKeyUp={onKeyUp}
          onKeyDown={onKeyDown}
        >
          Disabled Span
        </ButtonBase>,
      )
      const span = getByRole('button')

      fireEvent.click(span)
      expect(onClick).not.toHaveBeenCalled()

      fireEvent.keyUp(span)
      expect(onKeyUp).not.toHaveBeenCalled()

      fireEvent.keyDown(span)
      expect(onKeyDown).not.toHaveBeenCalled()
    })
  })
})
