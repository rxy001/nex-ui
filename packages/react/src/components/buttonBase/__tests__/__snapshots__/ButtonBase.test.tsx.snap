// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonBase should render with default props 1`] = `
<button
  tabindex="0"
  type="button"
>
  Button
</button>
`;
